### ISSUE TITLE
Union typed generics causes error of `cannot instantiate: 'T'`

### ISSUE THREAD

**Issue by binhonglee** [nim-lang/Nim#16153]
Function `echo` outputs the wrong string.

### Example
```nim
type
  X*[T] = openArray[T] | varargs[T] | seq[T]
  Y*[T] = ref object
    dll: T

proc initY*[T](): Y[T] =
  new(result)

proc initY*[T](v: X[T]): Y[T] =
  new(result)
  for e in v:
    echo e

var deque: Y[int] = initY[int]()
```

https://play.nim-lang.org/#ix=2FBN

### Current Output
```
Hint: used config file '/playground/nim/config/nim.cfg' [Conf]
Hint: used config file '/playground/nim/config/config.nims' [Conf]
....
/usercode/in.nim(14, 26) template/generic instantiation of `initY` from here
/usercode/in.nim(11, 12) template/generic instantiation of `items` from here
/playground/nim/lib/system/iterators.nim(6, 17) Error: cannot instantiate: 'T'
```

### Expected Output
(Basically success compilation.)
```
Hint: used config file '/playground/nim/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: in [Processing]
/usercode/in.nim(14, 5) Hint: 'deque' is declared but not used [XDeclaredButNotUsed]
Hint: gcc -c  -w   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/stdlib_system.nim.c.o /usercode/nimcache/stdlib_system.nim.c [Exec]
Hint: gcc -c  -w   -I/playground/nim/lib -I/usercode -o /usercode/nimcache/@min.nim.c.o /usercode/nimcache/@min.nim.c [Exec]
Hint:  [Link]
Hint: 14312 LOC; 1.815 sec; 15.934MiB peakmem; Debug build; proj: /usercode/in.nim; out: /usercode/in [SuccessX]
```

### Additional Information
This is a regression that worked fine on all the previous versions before 1.4.0 (since 0.19.0 according to playground).

**Comment by bung87** [comment]
it goes into `iterator instantiateGenericParamList`  `if q.typ.kind != tyCompositeTypeClass:` which is `tyGenericParam`, even it pass here it will error `type mismatch: got <openArray[T: not char]>` which mismatch to `func len*[TOpenArray: openArray|varargs](x: TOpenArray): int {.magic: "LengthOpenArray".} `

well, the root cause is `proc explicitGenericInstantiation` `a.kind` is `nkClosedSymChoice` it try instantiate two procs.

**Comment by ringabout** [comment]
@github-actions nim c
```nim
type
  X*[T] = openArray[T] | varargs[T] | seq[T]
  Y*[T] = ref object
    dll: T

proc initY*[T](): Y[T] =
  new(result)

proc initY*[T](v: X[T]): Y[T] =
  new(result)
  for e in v:
    echo e

var deque: Y[int] = initY[int]()
```

**Comment by ringabout** [comment]
!nim c
```nim
type
  X*[T] = openArray[T] | varargs[T] | seq[T]
  Y*[T] = ref object
    dll: T

proc initY*[T](): Y[T] =
  new(result)

proc initY*[T](v: X[T]): Y[T] =
  new(result)
  for e in v:
    echo e

var deque: Y[int] = initY[int]()
```

**Comment by metagn** [comment]
Works now, guessing fixed by #24010?

**Comment by ringabout** [comment]
!nim c
```nim
let code = """
type
  X*[T] = openArray[T] | varargs[T] | seq[T]
  Y*[T] = ref object
    dll: T

proc initY*[T](): Y[T] =
  new(result)

proc initY*[T](v: X[T]): Y[T] =
  new(result)
  for e in v:
    echo e

var deque: Y[int] = initY[int]()
"""

# foo()
writeFile("test.nim", code)

import std/osproc

let (_, exitCode) = execCmdEx("nim c -r test.nim")
doAssert exitCode != 0
```
